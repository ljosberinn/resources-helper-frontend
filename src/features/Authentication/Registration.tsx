import { Button, Control, Field, Help, Input, Label, Message } from 'rbx';
import React, { ChangeEvent, useCallback, FormEvent, useReducer } from 'react';
import { DebounceInput } from 'react-debounce-input';
import {
  isValidMail,
  isValidPassword,
  passwordPattern,
  ValidityIconLeft,
  ValidityIconRight,
} from '../../components/shared';
import useStoreon from 'storeon/react';
import { AuthenticationJSON } from '../../types';
import { History } from 'history';

const initialState = {
  mail: '',
  password: '',
  repeatedPassword: '',
  isSubmitting: false,
  mailError: '',
  passwordError: '',
  submitError: '',
};

type InitialStateType = typeof initialState;

type ReducerAction =
  | { type: 'SET_MAIL_ERROR'; value: string }
  | { type: 'SET_PASSWORD_ERROR'; value: string }
  | { type: 'SET_PASSWORD'; value: string }
  | { type: 'SET_MAIL'; value: string }
  | { type: 'SET_REPEATED_PASSWORD'; value: string }
  | { type: 'SET_IS_SUBMITTING'; value: boolean }
  | { type: 'SET_SUBMIT_ERROR'; value: string };

const reducer = (state: InitialStateType, action: ReducerAction) => {
  switch (action.type) {
    case 'SET_MAIL':
      return { ...state, mail: action.value };
    case 'SET_PASSWORD':
      return { ...state, password: action.value };
    case 'SET_REPEATED_PASSWORD':
      return { ...state, repeatedPassword: action.value };
    case 'SET_IS_SUBMITTING':
      return { ...state, isSubmitting: action.value };
    case 'SET_MAIL_ERROR':
      return { ...state, mailError: action.value };
    case 'SET_PASSWORD_ERROR':
      return { ...state, passwordError: action.value };
    case 'SET_SUBMIT_ERROR':
      return { ...state, submitError: action.value };
  }

  return state;
};

interface RegistrationProps {
  history: History;
}

export const Registration = ({ history }: RegistrationProps) => {
  const { dispatch } = useStoreon();

  const [
    {
      isSubmitting,
      mail,
      password,
      repeatedPassword,
      mailError,
      passwordError,
      submitError,
    },
    setState,
  ] = useReducer(reducer, initialState);

  const handleSetMail = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      const { value } = e.target;

      if (value !== mail) {
        setState({ type: 'SET_MAIL', value });
      }

      if (!isValidMail(value)) {
        setState({ type: 'SET_MAIL_ERROR', value: 'Invalid mail.' });
        return;
      }

      if (mailError.length > 0) {
        setState({ type: 'SET_MAIL_ERROR', value: '' });
      }
    },
    [mail, mailError],
  );

  const handleSetPassword = useCallback(
    (e: ChangeEvent<HTMLInputElement>, isRepeatedPassword = false) => {
      const { value } = e.target;

      setState({
        type: isRepeatedPassword ? 'SET_REPEATED_PASSWORD' : 'SET_PASSWORD',
        value,
      });

      if (!isValidPassword(value)) {
        setState({
          type: 'SET_PASSWORD_ERROR',
          value: 'Password not matching required pattern.',
        });
        return;
      }

      if (
        (isRepeatedPassword && value !== password) ||
        (!isRepeatedPassword && value !== repeatedPassword)
      ) {
        setState({
          type: 'SET_PASSWORD_ERROR',
          value: 'Passwords not matching.',
        });
        return;
      }

      if (passwordError.length > 0) {
        setState({ type: 'SET_PASSWORD_ERROR', value: '' });
      }
    },
    [password, repeatedPassword, passwordError],
  );

  const handleSubmit = useCallback(
    async (e: FormEvent<HTMLFormElement>) => {
      e.preventDefault();

      setState({
        type: 'SET_IS_SUBMITTING',
        value: true,
      });

      const body = new FormData();
      body.append('mail', mail);
      body.append('password', password);

      try {
        const response = await fetch('/register', { method: 'POST', body });
        const json = (await response.json()) as AuthenticationJSON;

        if (json.error) {
          switch (json.error) {
            case 'MAIL_EXISTS':
              setState({
                type: 'SET_SUBMIT_ERROR',
                value: 'Mail already exists.',
              });
              break;
            case 'PASSWORD_INSECURE':
              setState({
                type: 'SET_SUBMIT_ERROR',
                value: 'Provided password is considered insecure.',
              });
              break;
            default:
              setState({
                type: 'SET_SUBMIT_ERROR',
                value: json.error,
              });
          }

          setState({
            type: 'SET_IS_SUBMITTING',
            value: false,
          });

          return;
        }

        const { token } = json;

        dispatch('user/login', {
          token,
          isAuthenticated: true,
        });

        history.push('/');
      } catch (e) {
        setState({
          type: 'SET_SUBMIT_ERROR',
          value: 'Oops, something went wrong',
        });

        setState({
          type: 'SET_IS_SUBMITTING',
          value: false,
        });
      }
    },
    [mail, password, dispatch, history],
  );

  const maySubmit =
    !isValidMail(mail) ||
    mailError.length > 0 ||
    passwordError.length > 0 ||
    !isValidPassword(password) ||
    password.length === 0 ||
    repeatedPassword.length === 0 ||
    password !== repeatedPassword ||
    isSubmitting;

  return (
    <form onSubmit={handleSubmit}>
      <Message color="info">
        <Message.Header>Information</Message.Header>
        <Message.Body>
          Although <em>not necessary</em>, it is recommended to use a real mail
          for resetting forgotten credentials.
          <br />
          If you choose to select a real mail, you'll receive a mail containing
          your login details.
        </Message.Body>
      </Message>

      <Field grouped>
        <Field.Body>
          <Field>
            <Label>
              Mail
              <Control iconLeft iconRight>
                <Input
                  as={DebounceInput}
                  type="email"
                  onChange={handleSetMail}
                  size="small"
                  debounceTimeout={300}
                />
                <ValidityIconLeft type="mail" value={mail} />
                <ValidityIconRight type="mail" value={mail} />
              </Control>
            </Label>
            {mailError.length > 0 && <Help color="danger">{mailError}</Help>}
          </Field>

          <Field>
            <Label>
              Password
              <Control iconLeft iconRight>
                <Input
                  as={DebounceInput}
                  type="password"
                  onChange={handleSetPassword}
                  pattern={passwordPattern}
                  size="small"
                  debounceTimeout={300}
                />
                <ValidityIconLeft type="password" value={password} />
                <ValidityIconRight type="password" value={password} />
              </Control>
            </Label>
          </Field>

          <Field>
            <Label>
              Repeat password
              <Control iconLeft iconRight>
                <Input
                  as={DebounceInput}
                  type="password"
                  onChange={(e: ChangeEvent<HTMLInputElement>) =>
                    handleSetPassword(e, true)
                  }
                  pattern={passwordPattern}
                  size="small"
                  debounceTimeout={300}
                />
                <ValidityIconLeft type="password" value={repeatedPassword} />
                <ValidityIconRight type="password" value={repeatedPassword} />
              </Control>
            </Label>
            {passwordError.length > 0 && (
              <Help color="danger">{passwordError}</Help>
            )}
          </Field>
        </Field.Body>
      </Field>

      <Button
        type="submit"
        color="primary"
        disabled={maySubmit}
        state={isSubmitting ? 'loading' : undefined}
      >
        Submit
      </Button>
      {submitError.length > 0 && <Help color="danger">{submitError}</Help>}
    </form>
  );
};
