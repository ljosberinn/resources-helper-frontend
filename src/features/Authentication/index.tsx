import React from 'react';
import { withRouter } from 'react-router-dom';
import { Container, Column, Divider } from 'rbx';

import { Login } from './Login';
import { Registration } from './Registration';
import { History } from 'history';
import useStoreon from 'storeon/react';

interface AccountProps {
  history: History;
}

const Authentication = ({ history }: AccountProps) => {
  const { dispatch } = useStoreon();

  const props = { dispatch, history };

  return (
    <Container>
      <Column.Group centered>
        <Column size="one-third" textAlign="centered">
          <Login {...props} />
        </Column>
      </Column.Group>

      <div className="block">
        <Divider color="primary">OR</Divider>
      </div>

      <Column.Group centered>
        <Column size="two-thirds" textAlign="centered">
          <Registration {...props} />
        </Column>
      </Column.Group>
    </Container>
  );
};

export default withRouter(Authentication);
