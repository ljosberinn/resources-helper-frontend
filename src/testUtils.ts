const createValueEvent = (value: string) => ({ target: { value } });

export { createValueEvent };
