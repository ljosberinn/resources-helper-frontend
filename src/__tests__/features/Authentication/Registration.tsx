import React from 'react';
import ReactDOM from 'react-dom';
import { Registration } from '../../../features/Authentication/Registration';
import { createMemoryHistory } from 'history';
import {
  render,
  fireEvent,
  cleanup,
  waitForElement,
} from '@testing-library/react';
import { createValueEvent } from '../../../testUtils';
import axiosMock from 'axios';

afterEach(cleanup);

const INVALID_MAIL = 'abc';
const SHORT_PASSWORD = INVALID_MAIL;
const INVALID_PASSWORD = 'somelongerpassword';
const VALID_MAIL = 'a@b.de';
const VALID_PASSWORD = 'somePassword1!';

const INVALID_ICON_SELECTOR = 'span.icon.has-text-danger';
const VALID_ICON_SELECTOR = 'span.icon.has-text-success';
const WARNING_SELECTOR = 'p.help.is-danger';

const props = {
  dispatch: jest.fn(),
  history: createMemoryHistory(),
};

describe('<Registration />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');

    ReactDOM.render(<Registration {...props} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should display errors on invalid password repetition', async () => {
    // Arrange
    const { container } = render(<Registration {...props} />);
    const passwordInput = container.querySelector(
      '#register-2',
    ) as HTMLInputElement;

    const passwordRepetitionInput = container.querySelector(
      '#register-3',
    ) as HTMLInputElement;

    const passwordParent = passwordInput.parentElement as HTMLDivElement;
    const passwordRepetitionParent = passwordRepetitionInput.parentElement as HTMLDivElement;
    const submitButton = container.querySelector('button');

    // Act
    fireEvent.change(passwordInput, createValueEvent(INVALID_PASSWORD));
    await waitForElement(
      () => passwordParent.querySelectorAll(INVALID_ICON_SELECTOR).length > 0,
    );

    fireEvent.change(passwordRepetitionInput, createValueEvent(SHORT_PASSWORD));
    await waitForElement(
      () =>
        passwordRepetitionParent.querySelectorAll(INVALID_ICON_SELECTOR)
          .length > 0,
    );

    expect(passwordParent.querySelectorAll(INVALID_ICON_SELECTOR).length).toBe(
      2,
    );
    expect(passwordRepetitionParent.querySelector(WARNING_SELECTOR)).not.toBe(
      null,
    );

    expect(submitButton).toHaveProperty('disabled', true);
    expect(
      passwordInput.outerHTML + passwordRepetitionInput.outerHTML,
    ).toMatchSnapshot();
  });

  it('should display errors on valid mail and invalid password pattern', async () => {
    // Arrange
    const { container } = render(<Registration {...props} />);
    const mailInput = container.querySelector(
      '#register-1',
    ) as HTMLInputElement;
    const passwordInput = container.querySelector(
      '#register-2',
    ) as HTMLInputElement;
    const passwordRepetition = container.querySelector(
      '#register-3',
    ) as HTMLInputElement;
    const mailParent = mailInput.parentElement as HTMLDivElement;
    const passwordParent = passwordInput.parentElement as HTMLDivElement;
    const passwordRepetitionParent = passwordRepetition.parentElement as HTMLDivElement;
    const submitButton = container.querySelector('button');

    // Act
    fireEvent.change(mailInput, createValueEvent(VALID_MAIL));
    await waitForElement(
      () => mailParent.querySelectorAll(VALID_ICON_SELECTOR).length > 0,
    );

    fireEvent.change(passwordInput, createValueEvent(INVALID_PASSWORD));
    await waitForElement(
      () => passwordParent.querySelectorAll(INVALID_ICON_SELECTOR).length > 0,
    );

    // Assert
    expect(passwordParent.querySelectorAll(INVALID_ICON_SELECTOR).length).toBe(
      2,
    );
    expect(passwordRepetitionParent.querySelector(WARNING_SELECTOR)).not.toBe(
      null,
    );

    expect(submitButton).toHaveProperty('disabled', true);
    expect(mailInput.outerHTML + passwordInput.outerHTML).toMatchSnapshot();
  });

  it('should display errors on valid mail and short password', async () => {
    // Arrange
    const { container } = render(<Registration {...props} />);
    const mailInput = container.querySelector(
      '#register-1',
    ) as HTMLInputElement;
    const passwordInput = container.querySelector(
      '#register-2',
    ) as HTMLInputElement;
    const passwordRepetition = container.querySelector(
      '#register-3',
    ) as HTMLInputElement;
    const mailParent = mailInput.parentElement as HTMLDivElement;
    const passwordParent = passwordInput.parentElement as HTMLDivElement;
    const passwordRepetitionParent = passwordRepetition.parentElement as HTMLDivElement;
    const submitButton = container.querySelector('button');

    // Act
    fireEvent.change(mailInput, createValueEvent(VALID_MAIL));
    await waitForElement(
      () => mailParent.querySelectorAll(VALID_ICON_SELECTOR).length > 0,
    );

    fireEvent.change(passwordInput, createValueEvent(SHORT_PASSWORD));
    await waitForElement(
      () => passwordParent.querySelectorAll(INVALID_ICON_SELECTOR).length > 0,
    );

    // Assert
    expect(passwordParent.querySelectorAll(INVALID_ICON_SELECTOR).length).toBe(
      2,
    );
    expect(passwordRepetitionParent.querySelector(WARNING_SELECTOR)).not.toBe(
      null,
    );

    expect(submitButton).toHaveProperty('disabled', true);
    expect(mailParent.outerHTML + passwordParent.outerHTML).toMatchSnapshot();
  });

  it('should display errors on invalid mail and valid password pattern', async () => {
    // Arrange
    const { container } = render(<Registration {...props} />);
    const mailInput = container.querySelector(
      '#register-1',
    ) as HTMLInputElement;
    const passwordInput = container.querySelector(
      '#register-2',
    ) as HTMLInputElement;
    const mailParent = mailInput.parentElement as HTMLDivElement;
    const passwordParent = passwordInput.parentElement as HTMLDivElement;
    const submitButton = container.querySelector('button');

    // Act
    fireEvent.change(mailInput, createValueEvent(INVALID_MAIL));
    await waitForElement(
      () => mailParent.querySelectorAll(INVALID_ICON_SELECTOR).length > 0,
    );

    fireEvent.change(passwordInput, createValueEvent(VALID_PASSWORD));
    await waitForElement(
      () => passwordParent.querySelectorAll(VALID_ICON_SELECTOR).length > 0,
    );

    // Assert
    expect(mailParent.querySelectorAll(INVALID_ICON_SELECTOR).length).toBe(2);

    expect(mailParent.querySelector(WARNING_SELECTOR)).not.toBe(null);
    expect(passwordParent.querySelector(WARNING_SELECTOR)).toBe(null);

    expect(submitButton).toHaveProperty('disabled', true);
    expect(mailInput.outerHTML + passwordInput.outerHTML).toMatchSnapshot();
  });

  it('should submit with valid mail and valid password', async () => {
    // Arrange
    const { container } = render(<Registration {...props} />);
    const mailInput = container.querySelector(
      '#register-1',
    ) as HTMLInputElement;
    const passwordInput = container.querySelector(
      '#register-2',
    ) as HTMLInputElement;
    const passwordRepetitionInput = container.querySelector(
      '#register-3',
    ) as HTMLInputElement;
    const mailParent = mailInput.parentElement as HTMLDivElement;
    const submitButton = container.querySelector('button') as HTMLButtonElement;

    const spy = jest
      .spyOn(axiosMock, 'post')
      .mockImplementationOnce(() =>
        Promise.resolve({ data: { token: 'JWT' } }),
      );

    // Act
    fireEvent.change(mailInput, createValueEvent(VALID_MAIL));
    await waitForElement(
      () => mailParent.querySelectorAll(VALID_ICON_SELECTOR).length > 0,
    );

    fireEvent.change(passwordInput, createValueEvent(VALID_PASSWORD));
    await waitForElement(
      () => mailParent.querySelectorAll(VALID_ICON_SELECTOR).length > 0,
    );

    fireEvent.change(passwordRepetitionInput, createValueEvent(VALID_PASSWORD));
    await waitForElement(
      () => mailParent.querySelectorAll(VALID_ICON_SELECTOR).length > 0,
    );

    await waitForElement(() => submitButton.disabled === false);
    fireEvent.click(submitButton);

    // Assert
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(
      '/auth/register',
      {
        mail: VALID_MAIL,
        password: VALID_PASSWORD,
      },
      { timeout: 10000 },
    );
    expect(submitButton.classList.contains('is-loading')).toBeTruthy();
    expect(submitButton).toHaveProperty('disabled', true);
    expect(
      mailInput.outerHTML +
        passwordInput.outerHTML +
        passwordRepetitionInput.outerHTML,
    ).toMatchSnapshot();
  });
});
