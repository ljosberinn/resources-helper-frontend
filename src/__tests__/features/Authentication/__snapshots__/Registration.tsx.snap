// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Registration /> renders without crashing 1`] = `
<div>
  <form>
    <h1
      class="title"
    >
      Registration
    </h1>
    <article
      class="message is-info"
    >
      <div
        class="message-header"
      >
        Information
      </div>
      <div
        class="message-body"
      >
        Although 
        <em>
          not necessary
        </em>
        , it is recommended to use a real mail for resetting forgotten credentials.
        <br />
        If you choose to select a real mail, you'll receive a mail containing your login details.
      </div>
    </article>
    <div
      class="columns is-multiline"
    >
      <div
        class="column is-one-fifth is-offset-one-fifth"
      >
        <div
          class="field-label"
        >
          <label
            class="label"
            for="register-1"
          >
            Mail
          </label>
        </div>
      </div>
      <div
        class="column is-two-fifths"
      >
        <div
          class="field"
        >
          <div
            class="control has-icons-left has-icons-right"
          >
            <input
              class="input is-small"
              id="register-1"
              required=""
              type="email"
              value=""
            />
            <span
              class="icon is-left"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-envelope fa-w-16 "
                data-icon="envelope"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M502.3 190.8c3.9-3.1 9.7-.2 9.7 4.7V400c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V195.6c0-5 5.7-7.8 9.7-4.7 22.4 17.4 52.1 39.5 154.1 113.6 21.1 15.4 56.7 47.8 92.2 47.6 35.7.3 72-32.8 92.3-47.6 102-74.1 131.6-96.3 154-113.7zM256 320c23.2.4 56.6-29.2 73.4-41.4 132.7-96.3 142.8-104.7 173.4-128.7 5.8-4.5 9.2-11.5 9.2-18.9v-19c0-26.5-21.5-48-48-48H48C21.5 64 0 85.5 0 112v19c0 7.4 3.4 14.3 9.2 18.9 30.6 23.9 40.7 32.4 173.4 128.7 16.8 12.2 50.2 41.8 73.4 41.4z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
      <div
        class="column is-one-fifth is-offset-one-fifth"
      >
        <div
          class="field-label"
        >
          <label
            class="label"
            for="register-2"
          >
            Password
          </label>
        </div>
      </div>
      <div
        class="column is-two-fifths"
      >
        <div
          class="field"
        >
          <div
            class="control has-icons-left has-icons-right"
          >
            <input
              class="input is-small"
              id="register-2"
              pattern="^.*(?=.{8,})(?=.*[a-zA-Z])(?=.*\\\\d)(?=.*[!#$%&? \\"]).*$"
              required=""
              type="password"
              value=""
            />
            <span
              class="icon is-left"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-lock-open fa-w-18 "
                data-icon="lock-open"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 576 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M423.5 0C339.5.3 272 69.5 272 153.5V224H48c-26.5 0-48 21.5-48 48v192c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V272c0-26.5-21.5-48-48-48h-48v-71.1c0-39.6 31.7-72.5 71.3-72.9 40-.4 72.7 32.1 72.7 72v80c0 13.3 10.7 24 24 24h32c13.3 0 24-10.7 24-24v-80C576 68 507.5-.3 423.5 0z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
      <div
        class="column is-one-fifth is-offset-one-fifth"
      >
        <div
          class="field-label"
        >
          <label
            class="label"
            for="register-3"
          >
            Repeat password
          </label>
        </div>
      </div>
      <div
        class="column is-two-fifths"
      >
        <div
          class="field"
        >
          <div
            class="control has-icons-left has-icons-right"
          >
            <input
              class="input is-small"
              id="register-3"
              pattern="^.*(?=.{8,})(?=.*[a-zA-Z])(?=.*\\\\d)(?=.*[!#$%&? \\"]).*$"
              required=""
              type="password"
              value=""
            />
            <span
              class="icon is-left"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-lock-open fa-w-18 "
                data-icon="lock-open"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 576 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M423.5 0C339.5.3 272 69.5 272 153.5V224H48c-26.5 0-48 21.5-48 48v192c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V272c0-26.5-21.5-48-48-48h-48v-71.1c0-39.6 31.7-72.5 71.3-72.9 40-.4 72.7 32.1 72.7 72v80c0 13.3 10.7 24 24 24h32c13.3 0 24-10.7 24-24v-80C576 68 507.5-.3 423.5 0z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
    </div>
    <button
      class="button is-primary"
      disabled=""
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`<Registration /> should display errors on invalid mail and valid password pattern 1`] = `"<input class=\\"input is-small\\" type=\\"email\\" required=\\"\\" id=\\"register-1\\" value=\\"abc\\"><input class=\\"input is-small\\" type=\\"password\\" pattern=\\"^.*(?=.{8,})(?=.*[a-zA-Z])(?=.*\\\\d)(?=.*[!#$%&amp;? &quot;]).*$\\" required=\\"\\" id=\\"register-2\\" value=\\"somePassword1!\\">"`;

exports[`<Registration /> should display errors on invalid password repetition 1`] = `"<input class=\\"input is-small\\" type=\\"password\\" pattern=\\"^.*(?=.{8,})(?=.*[a-zA-Z])(?=.*\\\\d)(?=.*[!#$%&amp;? &quot;]).*$\\" required=\\"\\" id=\\"register-2\\" value=\\"somelongerpassword\\"><input class=\\"input is-small\\" type=\\"password\\" pattern=\\"^.*(?=.{8,})(?=.*[a-zA-Z])(?=.*\\\\d)(?=.*[!#$%&amp;? &quot;]).*$\\" required=\\"\\" id=\\"register-3\\" value=\\"abc\\">"`;

exports[`<Registration /> should display errors on valid mail and invalid password pattern 1`] = `"<input class=\\"input is-small\\" type=\\"email\\" required=\\"\\" id=\\"register-1\\" value=\\"a@b.de\\"><input class=\\"input is-small\\" type=\\"password\\" pattern=\\"^.*(?=.{8,})(?=.*[a-zA-Z])(?=.*\\\\d)(?=.*[!#$%&amp;? &quot;]).*$\\" required=\\"\\" id=\\"register-2\\" value=\\"somelongerpassword\\">"`;

exports[`<Registration /> should display errors on valid mail and short password 1`] = `"<div class=\\"control has-icons-left has-icons-right\\"><input class=\\"input is-small\\" type=\\"email\\" required=\\"\\" id=\\"register-1\\" value=\\"a@b.de\\"><span class=\\"icon has-text-success is-left\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"envelope\\" class=\\"svg-inline--fa fa-envelope fa-w-16 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M502.3 190.8c3.9-3.1 9.7-.2 9.7 4.7V400c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V195.6c0-5 5.7-7.8 9.7-4.7 22.4 17.4 52.1 39.5 154.1 113.6 21.1 15.4 56.7 47.8 92.2 47.6 35.7.3 72-32.8 92.3-47.6 102-74.1 131.6-96.3 154-113.7zM256 320c23.2.4 56.6-29.2 73.4-41.4 132.7-96.3 142.8-104.7 173.4-128.7 5.8-4.5 9.2-11.5 9.2-18.9v-19c0-26.5-21.5-48-48-48H48C21.5 64 0 85.5 0 112v19c0 7.4 3.4 14.3 9.2 18.9 30.6 23.9 40.7 32.4 173.4 128.7 16.8 12.2 50.2 41.8 73.4 41.4z\\"></path></svg></span><span class=\\"icon has-text-success is-right\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"check\\" class=\\"svg-inline--fa fa-check fa-w-16 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z\\"></path></svg></span></div><div class=\\"control has-icons-left has-icons-right\\"><input class=\\"input is-small\\" type=\\"password\\" pattern=\\"^.*(?=.{8,})(?=.*[a-zA-Z])(?=.*\\\\d)(?=.*[!#$%&amp;? &quot;]).*$\\" required=\\"\\" id=\\"register-2\\" value=\\"abc\\"><span class=\\"icon has-text-danger is-left\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"lock-open\\" class=\\"svg-inline--fa fa-lock-open fa-w-18 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 576 512\\"><path fill=\\"currentColor\\" d=\\"M423.5 0C339.5.3 272 69.5 272 153.5V224H48c-26.5 0-48 21.5-48 48v192c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V272c0-26.5-21.5-48-48-48h-48v-71.1c0-39.6 31.7-72.5 71.3-72.9 40-.4 72.7 32.1 72.7 72v80c0 13.3 10.7 24 24 24h32c13.3 0 24-10.7 24-24v-80C576 68 507.5-.3 423.5 0z\\"></path></svg></span><span class=\\"icon has-text-danger is-right\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"times\\" class=\\"svg-inline--fa fa-times fa-w-11 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 352 512\\"><path fill=\\"currentColor\\" d=\\"M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z\\"></path></svg></span></div>"`;

exports[`<Registration /> should submit with valid mail and valid password 1`] = `"<input class=\\"input is-small\\" type=\\"email\\" required=\\"\\" id=\\"register-1\\" value=\\"a@b.de\\" disabled=\\"\\"><input class=\\"input is-small\\" type=\\"password\\" pattern=\\"^.*(?=.{8,})(?=.*[a-zA-Z])(?=.*\\\\d)(?=.*[!#$%&amp;? &quot;]).*$\\" required=\\"\\" id=\\"register-2\\" value=\\"somePassword1!\\" disabled=\\"\\"><input class=\\"input is-small\\" type=\\"password\\" pattern=\\"^.*(?=.{8,})(?=.*[a-zA-Z])(?=.*\\\\d)(?=.*[!#$%&amp;? &quot;]).*$\\" required=\\"\\" id=\\"register-3\\" value=\\"somePassword1!\\">"`;
